src/circular_buffer.cpp

```diff
@@ -0,0 +1,80 @@
+#include "circular_buffer.h"
+
+#include <algorithm>
+#include <cassert>
+
+namespace lightstep {
+//------------------------------------------------------------------------------
+// constructor
+//------------------------------------------------------------------------------
+CircularBuffer::CircularBuffer(size_t max_size)
+    : data_{new char[max_size + 1]}, capacity_{max_size + 1} {}
+
+//------------------------------------------------------------------------------
+// Reserve
+//------------------------------------------------------------------------------
+CircularBufferPlacement CircularBuffer::Reserve(size_t num_bytes) noexcept {
+  size_t new_head;
+  size_t head = head_;
+  while (true) {
+    size_t tail = tail_;
+
+    // Because we update head_ before tail_, it's possible that tail > head; if
```

Comment:
Can you confirm: this might happen when due to a race other threads have both updated head_ and tail_. IOW if (tail > head) it means both have been updated by someone else?

Reply:
Yes, the tail would get updated by the consumer thread (there can be only one) and the head would get updated by another producer thread (no limitation on producers).